### PR TITLE
[BIOMAGE-2185] - Improve `DotPlot` work response

### PR DESCRIPTION
**Pull request by aerlaut**
# Description
Improve DotPlot work response

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-2185

#### Link to staging deployment URL (or set N/A)
N/A

#### Links to any PRs or resources related to this PR
UI - https://github.com/biomage-org/ui/pull/80

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: r/R/dotplot.R [hunk 1]
```diff
@@ -71,17 +71,28 @@ runDotPlot <- function(req, data) {
...
   # features.plot has the ensemble ids: get gene symbols
   dotplot_data$name <- features[dotplot_data$features.plot, "name"]
   dotplot_data <- dotplot_data[stringr::str_order(dotplot_data$id, numeric = TRUE), ]
+
   dotplot_data <- dotplot_data %>%
     dplyr::transmute(
-      cellSets = as.character(id),
-      geneName = as.character(name),
+      cellSetsIdx = id,
+      geneNameIdx = factor(name),
```
Comment (gerbeldo) [inline r/R/dotplot.R:82]: You should be careful when using `factor` and `levels` on `character` variables. In this case I think it's OK, but is nice to know, as it is a common gotcha.

Factors order string levels *alphabetically*. When you take the `levels(dotplot_data$geneNameIdx)` below, the order might not be what you expect it to be.

File: r/tests/testthat/_snaps/dotplot.md [hunk 2]
```diff
@@ -3,284 +3,28 @@
     Code
       res
     Output
-      [[1]]
-      [[1]]$cellSets
-      [1] "hello"
+      $cellSetsIdx
```
Comment (gerbeldo) [inline r/tests/testthat/_snaps/dotplot.md:6]: noice snapshot 👌

File: r/tests/testthat/test-dotplot.R [hunk 3]
```diff
@@ -96,7 +97,6 @@ test_that("Dotplot returns the correct values", {
 
   scaled_res <- scale(correct_res)
   scaled_res <- Seurat::MinMax(scaled_res, min = -2.5, max = 2.5)
-  scaled_res <- as.list(x = t(x = scaled_res))
-  dotPlot_res <- unname(purrr::transpose(dotPlot_res)$avgExpression)
-  expect_equal(dotPlot_res, scaled_res)
+  scaled_res <- unlist(as.list(x = t(x = scaled_res)))
```
Comment (gerbeldo) [inline r/tests/testthat/test-dotplot.R:100]: this is quite ugly I would say, but if it's the expected shape, and in a test, it's fine. scaled_res is some kind of table right (not a one dimensional vector or list?)

**Review by gerbeldo — APPROVED** [review]
looks good, left some comments

File: r/R/dotplot.R (hunk 1 above)
Comment (aerlaut) [reply r/R/dotplot.R:82]: Thanks for the heads up Ger! I didn't realize that `factor` sorts the levels alphabetically. I'm getting the internal representation of the factors and its level - I didn't mess around with the index. So it should be matching up the the names and the index properly right?

File: r/tests/testthat/_snaps/dotplot.md (hunk 2 above)
Comment (aerlaut) [reply r/tests/testthat/_snaps/dotplot.md:6]: image

File: r/tests/testthat/test-dotplot.R (hunk 3 above)
Comment (aerlaut) [reply r/tests/testthat/test-dotplot.R:100]: yes, it's a table, that's why we can transpose it.

File: r/R/dotplot.R (hunk 1 above)
Comment (gerbeldo) [reply r/R/dotplot.R:82]: exactly! factors are integers that index the levels. you can access the ints with `as.integer(your_factor)`. and yep, should be matching correctly 👍

**Pull request merged**